numba/array.py

```diff
@@ -0,0 +1,177 @@
+            self._sync('host')
+
+    def __array__(self, *args):
+
+        self._sync('host')
+        return numpy.array(self._host, *args)
+
+    def _sync(self, where):
+        """Sync the data in one memory space with the other."""
+
+        if where == 'gpu':
+            if self._gpu is None:
+                self._allocate('gpu', None, self.dtype, self.shape, self.strides)
+            if not self._gpu_valid:
+                self._copy_to_gpu()
+        else:
+            if self._host is None:
+                self._allocate('host', None, self.dtype, self.shape, self.strides)
+            if not self._host_valid:
+                self._copy_to_host()
+
+    @trace
+    def _invalidate(self, where):
+        """Mark the host / device array as out-of-date."""
+
+        if where == 'gpu':
+            self._gpu_valid = False
+        else:
+            self._host_valid = False
+
+    @trace
+    def _allocate(self, where, obj=None, dtype=None, shape=None, strides=None,
+                  copy=True):
+        if dtype:
+            dtype = numpy.dtype(dtype)
+        if where == 'host':
+            if obj is not None:
+                self._host = numpy.array(obj, dtype, copy=copy)
+            else:
+                self._host = numpy.empty(shape, dtype, _s2o(dtype, shape, strides))
+        else:
+            # Don't import this at module-scope as it may not be available
+            # in all environments (e.g., CUDASIM)
+            from numba.cuda.cudadrv import devicearray as da
+            if obj is not None:
+                # If 'obj' is an array-like object but not an ndarray,
+                # construct an ndarray first to extract all the parameters we need.
+                if not isinstance(obj, numpy.ndarray):
+                    obj = numpy.array(obj, copy=False)
+                self._gpu = da.from_array_like(obj)
+            else:
+                if strides is None:
+                    strides = _o2s(dtype, shape, 'C')
+                self._gpu = da.DeviceNDArray(shape, strides, dtype)
+
+    @trace
+    def _copy_to_gpu(self):
+        self._gpu.copy_to_device(self._host)
+        self._gpu_valid = True
+
+    @trace
+    def _copy_to_host(self):
+        self._gpu.copy_to_host(self._host)
+        self._host_valid = True
+
+    #def __hasattr__(self, attr):#
+
+    #    # forward all normal operations to the host array
+    #    return self.host().__hasattr__(attr)
+
+    @trace
+    def __getattr__(self, name):
+        """Transparently forward attribute access to the host array."""
+
+        if self._host is None:
+            self._allocate('host', None, self.dtype, self.shape, self.strides)
+
+        # FIXME: for some attributes we need to sync first !
+        return getattr(self._host, name)
+
+    def __len__(self, *args): return self.size
+    def __eq__(self, other):
+        if type(self) is not type(other): return False
+        # FIXME: If both arrays have valid GPU data, compare there.
```

Comment:
Well, in the current implementation that's done in the line below this comment (`self.host() == other.host()`). But as the comment above states, if both arrays are valid on the GPU, we may not have to sync to the host, if there is a suitable comparison operation defined in a CUDA kernel.